posts/nix-vs-docker-i.md

```diff
@@ -0,0 +1,260 @@
+
+```nix
+dockerTools.buildImage {
+  name = "myapp";
+  config = {
+    Cmd = [ "node" "index.js" ];
+  };
+  contents = [ myApp ];
+}
+```
+
+This gives you a reproducible Docker image---without relying on a Dockerfile or
+layers at all.
+
+## Security and Isolation
+
+Docker builds often run with elevated privileges. The Docker daemon runs as
+root, and most images execute shell commands during build. Unless you harden
+every stage, you're at risk. Sandboxing is optional. Escape vectors are known.
+
+Nix doesn't run a daemon. It doesn't need root. And by default, it builds in a
+sandbox with no network, no writable files, and no access to your home
+directory.
+
+Here's an example failure you'd get if you try to break the sandbox:
+
+```sh
+cp ~/.ssh/id_rsa .
+```
+
+If you're building with `--option sandbox true`, this fails. The sandbox
+prevents access outside declared paths.
+
+This model isn't just more secure. It's verifiable. You can trace the entire
+dependency graph and inspect every file that went into the build. That's the
+foundation NixOS builds on---here your entire OS, including the kernel, is just
+another derivation.
+
+## Tooling and Ecosystem
+
+Docker is supported by a wide array of third-party tools. Compose, BuildKit,
+Docker Hub, Portainer, etc. Each one of those tools is trying to solve a piece
+of the container puzzle, but these tools are fragmented and often orthogonal.
+Compose files don't integrate cleanly with Dockerfiles. Secrets management is
+bolted on. Native caching and layering are leaky abstractions at best.
+
+The Nix ecosystem, in contrast, is more unified. Flakes, despite their current
+state and perception by the community, give you a single entry point for
+dependency specification, building, testing, and deployment. You define your
+inputs, and then you define your outputs: your packages, formatters, or your
+devShells in one file, and they work across all tools that support the flake
+interface. In fact, you can even bypass the flake interface with a project that
+uses flakes _reproducibly_. It is that integrated.
+
+Need to test a system configuration? Use `nixos-rebuild test` with a flake. Need
+a temporary environment for hacking? `nix develop`. Want to run integration
+tests? Use `nixos-test`. Software tests? Nix has
+[got your back](https://notashelf.dev/posts/nixos-testing-i). Tools like
+`lorri`, `direnv`, and `nixd` enhance developer ergonomics with automatic
+environment loading and IDE integration.
+
+Instead of chaining loosely-coupled CLIs and YAML fragments, Nix encourages
+composition. Everything is just a function, and everything is declarative. This
+reduces surface area and makes behavior more predictable.
+
+## Closing Thoughts
+
+Docker solved the deployment problem for a world of mutable systems and
+inconsistent development environments. But it did so by building on top of an
+architecture that makes reproducibility an afterthought. It trades convenience
```

Comment:
Engrish hard, but I think this makes more sense: "It trades (gives away) long-term reliability for (to obtain) convenience."